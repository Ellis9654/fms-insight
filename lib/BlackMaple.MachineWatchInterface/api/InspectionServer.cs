using System.Collections.Generic;

namespace BlackMaple.MachineWatchInterface
{
    public interface IInspectionControl
    {
        //Forces the given materialID to be inspected
        void ForceInspection(long materialID, string inspectionType);

        //Forces the next piece that runs on the given location to be inspected
        void NextPieceInspection(MachineWatchInterface.PalletLocation palLoc, string inspType);

        //allow editing of counts.  Uses lists of InspectCount structure
        List<InspectCount> LoadInspectCounts();
        void SetInspectCounts(IEnumerable<InspectCount> countUpdates);

        List<string> LoadGlobalInspectionTypes();

        InspectionType LoadGlobalInspectionType(string ty);

        void SetGlobalInspectionType(InspectionType ty);

        void DeleteGlobalInspectionType(string ty);
    }
}
