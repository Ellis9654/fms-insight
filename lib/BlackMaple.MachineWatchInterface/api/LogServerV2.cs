using System;
using System.Collections.Generic;

namespace BlackMaple.MachineWatchInterface
{
    [Serializable]
    public enum SerialType
    {
        NoSerials,
        OneSerialPerMaterial,  // assign a different serial to each piece of material
        OneSerialPerCycle,     // assign a single serial to all the material on each cycle
        SerialDeposit          // deposit serial into machine file to be scribed on part
    }

    [Serializable]
    public class SerialSettings
    {
        public SerialType SerialType {get;}
        public int SerialLength {get;}

        //settings only for serial deposit
        public int DepositOnProcess {get;}
        public string FilenameTemplate {get;}
        public string ProgramTemplate {get;}

        public SerialSettings(SerialType t, int len)
        {
            SerialType = t;
            SerialLength = len;
            DepositOnProcess = 1;
            FilenameTemplate = null;
            ProgramTemplate = null;
        }
        public SerialSettings(int len, int proc, string fileTemplate, string progTemplate)
        {
            SerialType = SerialType.SerialDeposit;
            SerialLength = len;
            DepositOnProcess = proc;
            FilenameTemplate = fileTemplate;
            ProgramTemplate = progTemplate;
        }
    }

    public interface ILogServerV2
    {
        List<LogEntry> GetLogEntries(DateTime startUTC, DateTime endUTC);
        List<LogEntry> GetLog(long lastSeenCounter);
        List<LogEntry> GetLogForMaterial(long materialID);
        List<LogEntry> GetLogForSerial(string serial);
        List<LogEntry> GetLogForWorkorder(string workorder);
        List<LogEntry> GetCompletedPartLogs(DateTime startUTC, DateTime endUTC);
        List<WorkorderSummary> GetWorkorderSummaries(IEnumerable<string> workorderIds);

        LogEntry RecordSerialForMaterialID(LogMaterial mat, string serial);
        LogEntry RecordWorkorderForMaterialID(LogMaterial mat, string workorder);
        LogEntry RecordFinalizedWorkorder(string workorder);

        SerialSettings GetSerialSettings();
        void SetSerialSettings(SerialSettings s);
    }
}

