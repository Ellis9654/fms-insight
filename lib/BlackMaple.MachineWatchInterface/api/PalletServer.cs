using System;
using System.Collections.Generic;

namespace BlackMaple.MachineWatchInterface
{
    public interface IPalletServer
    {
        // Update pallet status
        void SetHold(string pal, bool hold);
        void MarkMaterialLoadedOntoPallet(string pal, string face, long matID);
        void MarkLoadUnloadInstructionsCompleted(string pal);
        void RejectLoadUnloadInstructions(string pal);
        void StartManualPalletMove(string pal, PalletLocation target);
        void ScrapMaterialFromPallet(string pal, int materialID);

        // Override pallet status
        void OverridePalletLocation(string pal, PalletLocation currentLocation);
        void OverrideFixtureOnPallet(string pal, string fixture);
    }
}
