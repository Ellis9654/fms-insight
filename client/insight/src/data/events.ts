/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import { addDays, addMonths, startOfMonth } from 'date-fns';
import { PledgeStatus, ConsumingPledge } from './pledge';
import * as im from 'immutable';

import * as api from './api';
import * as oee from './events.oee';
import * as cycles from './events.cycles';
import * as matsummary from './events.matsummary';

export { OeeState, StationInUse } from './events.oee';
export { CycleState, CycleData, binCyclesByDayAndStat, binCyclesByDayAndPart } from './events.cycles';
export { MaterialSummary } from './events.matsummary';

export enum AnalysisPeriod {
    Last30Days = 'Last_30_Days',
    SpecificMonth = 'Specific_Month'
}

export interface Last30Days {
    readonly latest_log_entry: Date | undefined;
    readonly latest_log_counter: number | undefined;

    readonly latest_scheduleId: string | undefined;

    readonly oee: oee.OeeState;
    readonly cycles: cycles.CycleState;
    readonly mat_summary: matsummary.MatSummaryState;
}

export interface AnalysisMonth {
    readonly cycles: cycles.CycleState;
}

const emptyAnalysisMonth: AnalysisMonth = {
    cycles: cycles.initial
};

export interface State {
    readonly loading_log_entries: boolean;
    readonly loading_job_history: boolean;
    readonly loading_error?: Error;

    readonly analysis_period: AnalysisPeriod;
    readonly analysis_period_month: Date;
    readonly loading_analysis_month_log: boolean;
    readonly loading_analysis_month_jobs: boolean;

    readonly last30: Last30Days;
    readonly selected_month: AnalysisMonth;
}

export const initial: State = {
    loading_log_entries: false,
    loading_job_history: false,
    analysis_period: AnalysisPeriod.Last30Days,
    analysis_period_month: startOfMonth(new Date()),
    loading_analysis_month_log: false,
    loading_analysis_month_jobs: false,

    last30: {
        latest_log_entry: undefined,
        latest_log_counter: undefined,
        latest_scheduleId: undefined,
        oee: oee.initial,
        cycles: cycles.initial,
        mat_summary: matsummary.initial,
    },

    selected_month: emptyAnalysisMonth,
};

export enum ActionType {
    SetAnalysisLast30Days = 'Events_SetAnalysisLast30Days',
    SetAnalysisMonth = 'Events_SetAnalysisMonth',
    LoadRecentLogEntries = 'Events_LoadRecentLogEntries',
    LoadRecentJobHistory = 'Events_LoadRecentJobHistory',
    LoadSpecificMonthLogEntries = 'Events_LoadSpecificMonthLogEntries',
    LoadSpecificMonthJobHistory = 'Events_LoadSpecificMonthJobHistory',
    SetSystemHours = 'Events_SetSystemHours',
    ReceiveNewLogEntries = 'Events_NewLogEntries',
    ReceiveNewJobs = 'Events_ReceiveNewJobs',
    Other = 'Other',
}

// TODO: use Plege when typescript 2.8 shows up
export type Action =
  | {type: ActionType.SetAnalysisLast30Days }
  | {type: ActionType.SetAnalysisMonth, month: Date }
  | {type: ActionType.LoadRecentLogEntries, now: Date, pledge: ConsumingPledge<ReadonlyArray<Readonly<api.ILogEntry>>>}
  | {type: ActionType.LoadRecentJobHistory, now: Date, pledge: ConsumingPledge<Readonly<api.IHistoricData>>}
  | {
      type: ActionType.LoadSpecificMonthLogEntries,
      month: Date,
      pledge: ConsumingPledge<ReadonlyArray<Readonly<api.ILogEntry>>>
    }
  | {
      type: ActionType.LoadSpecificMonthJobHistory,
      month: Date,
      pledge: ConsumingPledge<Readonly<api.IHistoricData>>
    }
  | {type: ActionType.ReceiveNewLogEntries, now: Date, events: ReadonlyArray<Readonly<api.ILogEntry>>}
  | {type: ActionType.ReceiveNewJobs, now: Date, jobs: Readonly<api.IHistoricData>}
  | {type: ActionType.SetSystemHours, hours: number}
  | {type: ActionType.Other}
  ;

export function loadLast30Days() /*: Action<ActionUse.CreatingAction> */ {
    var logClient = new api.LogClient();
    var jobClient = new api.JobsClient();
    var now = new Date();
    var thirtyDaysAgo = addDays(now, -30);
    return [
        {
            type: ActionType.LoadRecentLogEntries,
            now: now,
            pledge: logClient.get(thirtyDaysAgo, now)
        },
        {
            type: ActionType.LoadRecentJobHistory,
            now: now,
            pledge: jobClient.history(thirtyDaysAgo, now)

        }
    ];
}

export function refreshLogEntries(lastCounter: number) {
    var client = new api.LogClient();
    var now = new Date();
    return {
        type: ActionType.LoadRecentLogEntries,
        now: now,
        pledge: client.recent(lastCounter)
    };
}

export function receiveNewEvents(events: ReadonlyArray<Readonly<api.ILogEntry>>): Action {
    return {
        type: ActionType.ReceiveNewLogEntries, now: new Date(), events
    };
}

export function receiveNewJobs(newJobs: Readonly<api.INewJobs>): Action {
    const jobs: {[key: string]: api.JobPlan} = {};
    newJobs.jobs.forEach(j => {
        jobs[j.unique] = j;
    });
    return {
        type: ActionType.ReceiveNewJobs, now: new Date(), jobs: {
            jobs: jobs,
            stationUse: newJobs.stationUse
        }
    };
}

export function analyzeLast30Days() {
    return {type: ActionType.SetAnalysisLast30Days};
}

export function analyzeSpecificMonth(month: Date) {
    var client = new api.LogClient();
    var startOfNextMonth = addMonths(month, 1);
    return {
        type: ActionType.LoadSpecificMonthLogEntries,
        month: month,
        pledge: client.get(month, startOfNextMonth)
    };
}

export function setAnalysisMonth(month: Date) {
    return {
        type: ActionType.SetAnalysisMonth,
        month: month
    };
}

function safeAssign<T extends R, R>(o: T, n: R): T {
    let allMatch: boolean = true;
    for (let k of Object.keys(n)) {
        // tslint:disable-next-line:no-any
        if ((o as any)[k] !== (n as any)[k]) {
            allMatch = false;
            break;
        }
    }
    if (allMatch) {
        return o;
    } else {
        return Object.assign({}, o, n);
    }
}

function processRecentLogEntries(now: Date, evts: Iterable<api.ILogEntry>, s: Last30Days): Last30Days {
    var thirtyDaysAgo = addDays(now, -30);
    let lastCounter = s.latest_log_counter;
    let lastDate = s.latest_log_entry;
    let lastNewEvent = im.Seq(evts).maxBy(e => e.counter);
    if (lastNewEvent !== undefined) {
        if (lastCounter === undefined || lastCounter < lastNewEvent.counter) {
            lastCounter = lastNewEvent.counter;
            lastDate = lastNewEvent.endUTC;
        }
    }
    return safeAssign(
        s,
        {
            latest_log_counter: lastCounter,
            latest_log_entry: lastDate,
            oee: oee.process_events(now, evts, s.oee),
            cycles: cycles.process_events(
                {type: cycles.ExpireOldDataType.ExpireEarlierThan, d: thirtyDaysAgo},
                evts,
                s.cycles),
            mat_summary: matsummary.process_events(now, evts, s.mat_summary)
        });
}

function processSpecificMonthLogEntries(evts: Iterable<api.ILogEntry>, s: AnalysisMonth): AnalysisMonth {
    return safeAssign(
        s,
        {
            cycles: cycles.process_events(
                {type: cycles.ExpireOldDataType.NoExpire},
                evts,
                s.cycles)
        }
    );
}

function processRecentJobs(now: Date, jobs: Readonly<api.IHistoricData>, s: Last30Days): Last30Days {
    let latestSchId = s.latest_scheduleId;
    let largestJob = im.Seq(jobs.jobs).maxBy(j => j.scheduleId);
    if (largestJob !== undefined) {
        if (latestSchId === undefined || (largestJob.scheduleId !== undefined) && latestSchId < largestJob.scheduleId) {
            latestSchId = largestJob.scheduleId;
        }
    }
    return safeAssign(
        s,
        {
            latest_scheduleId: latestSchId
        }
    );
}

function processSpecificMonthJobs(jobs: Readonly<api.IHistoricData>, s: AnalysisMonth): AnalysisMonth {
    return s;
}

export function reducer(s: State, a: Action): State {
    if (s === undefined) { return initial; }
    switch (a.type) {
        case ActionType.LoadRecentLogEntries:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s, loading_log_entries: true, loading_error: undefined};
                case PledgeStatus.Completed:
                    return {...s,
                        last30: processRecentLogEntries(a.now, a.pledge.result, s.last30),
                        loading_log_entries: false
                    };
                case PledgeStatus.Error:
                    return {...s, loading_log_entries: false, loading_error: a.pledge.error};
                default: return s;
            }

        case ActionType.ReceiveNewLogEntries:
            return {...s,
                last30: processRecentLogEntries(a.now, a.events, s.last30)
            };

        case ActionType.LoadRecentJobHistory:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s, loading_job_history: true, loading_error: undefined};
                case PledgeStatus.Completed:
                    return {...s,
                        last30: processRecentJobs(a.now, a.pledge.result, s.last30),
                        loading_job_history: false
                    };
                case PledgeStatus.Error:
                    return {...s, loading_job_history: false, loading_error: a.pledge.error};
                default: return s;
            }

        case ActionType.ReceiveNewJobs:
            return {...s,
                last30: processRecentJobs(a.now, a.jobs, s.last30)
            };

        case ActionType.SetSystemHours:
            return {...s,
                last30: {...s.last30,
                    oee: {...s.last30.oee, system_active_hours_per_week: a.hours}
                },
            };

        case ActionType.SetAnalysisLast30Days:
            return {...s,
                analysis_period: AnalysisPeriod.Last30Days,
                selected_month: emptyAnalysisMonth,
            };

        case ActionType.LoadSpecificMonthLogEntries:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s,
                        analysis_period: AnalysisPeriod.SpecificMonth,
                        analysis_period_month: a.month,
                        loading_error: undefined,
                        loading_analysis_month_log: true,
                        selected_month: emptyAnalysisMonth,
                    };

                case PledgeStatus.Completed:
                    return {...s,
                        loading_analysis_month_log: false,
                        selected_month: processSpecificMonthLogEntries(a.pledge.result, s.selected_month)
                    };

                case PledgeStatus.Error:
                    return {...s,
                        loading_analysis_month_log: false,
                        loading_error: a.pledge.error
                    };

                default: return s;
            }

        case ActionType.LoadSpecificMonthJobHistory:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s,
                        loading_analysis_month_jobs: true,
                    };

                case PledgeStatus.Completed:
                    return {...s,
                        loading_analysis_month_jobs: false,
                        selected_month: processSpecificMonthJobs(a.pledge.result, s.selected_month)
                    };

                case PledgeStatus.Error:
                    return {...s,
                        loading_analysis_month_jobs: false,
                        loading_error: a.pledge.error
                    };

                default: return s;
            }

        case ActionType.SetAnalysisMonth:
            return {...s, analysis_period_month: a.month};

        default: return s;
    }
}