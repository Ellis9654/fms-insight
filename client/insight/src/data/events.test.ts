/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { addDays, addHours } from 'date-fns';
import { duration } from 'moment';

import { PledgeStatus } from './pledge';
import * as events from './events';
import * as stationCycles from './events.cycles';
import { fakeCycle } from './events.fake';
import { ILogEntry } from './api';

it('creates initial state', () => {
  // tslint:disable no-any
  let s = events.reducer(undefined as any, undefined as any);
  // tslint:enable no-any
  expect(s).toBe(events.initial);
});

it('sets the system hours', () => {
  let s = events.reducer(events.initial, {
    type: events.ActionType.SetSystemHours,
    hours: 5
  });
  expect(s.last30.oee.system_active_hours_per_week).toBe(5);
});

it('responds to loading', () => {
  let st = events.reducer(
    {...events.initial, loading_error: new Error('hello')},
    {
      type: events.ActionType.LoadRecentLogEntries,
      now: new Date(),
      pledge: {
        status: PledgeStatus.Starting
      }
    }
  );
  expect(st.loading_log_entries).toBe(true);
  expect(st.loading_job_history).toBe(false);
  expect(st.loading_error).toBeUndefined();
  expect(st.last30).toBe(events.initial.last30);
  expect(st.selected_month).toBe(events.initial.selected_month);
});

it('responds to loading for jobs', () => {
  let st = events.reducer(
    {...events.initial, loading_error: new Error('hello')},
    {
      type: events.ActionType.LoadRecentJobHistory,
      now: new Date(),
      pledge: {
        status: PledgeStatus.Starting
      }
    }
  );
  expect(st.loading_log_entries).toBe(false);
  expect(st.loading_job_history).toBe(true);
  expect(st.loading_error).toBeUndefined();
  expect(st.last30).toBe(events.initial.last30);
  expect(st.selected_month).toBe(events.initial.selected_month);
});

it('responds to error', () => {
  let st = events.reducer(
    {...events.initial, loading_log_entries: true},
    {
      type: events.ActionType.LoadRecentLogEntries,
      now: new Date(),
      pledge: {
        status: PledgeStatus.Error,
        error: new Error('hello')
      }
    }
  );
  expect(st.loading_log_entries).toBe(false);
  expect(st.loading_job_history).toBe(false);
  expect(st.loading_error).toEqual(new Error('hello'));
  expect(st.last30).toBe(events.initial.last30);
  expect(st.selected_month).toBe(events.initial.selected_month);
});

it('responds to error for jobs', () => {
  let st = events.reducer(
    {...events.initial, loading_job_history: true},
    {
      type: events.ActionType.LoadRecentJobHistory,
      now: new Date(),
      pledge: {
        status: PledgeStatus.Error,
        error: new Error('hello')
      }
    }
  );
  expect(st.loading_log_entries).toBe(false);
  expect(st.loading_job_history).toBe(false);
  expect(st.loading_error).toEqual(new Error('hello'));
  expect(st.last30).toBe(events.initial.last30);
  expect(st.selected_month).toBe(events.initial.selected_month);
});

function procNewEvents(evtsToAction: (now: Date, newEvts: ReadonlyArray<ILogEntry>) => events.Action) {
  var now = new Date(2018, 1, 2, 3, 4, 5);

  // start with cycles from 27 days ago, 2 days ago, and today
  const todayCycle = fakeCycle(now, 30, "part111", 1, 'palbb');
  var twoDaysAgo = addDays(now, -2);
  const twoDaysAgoCycle = fakeCycle(twoDaysAgo, 24, "part222", 1, 'palbb');
  var twentySevenDaysAgo = addDays(now, -27);
  const twentySevenCycle = fakeCycle(twentySevenDaysAgo, 18, "part222", 2, 'palaa');
  let st = events.reducer(
    events.initial,
    {
      type: events.ActionType.LoadRecentLogEntries,
      now: now,
      pledge: {
        status: PledgeStatus.Completed,
        result: twentySevenCycle.concat(twoDaysAgoCycle, todayCycle)
      }
    });
  expect(st.last30).toMatchSnapshot('last30 with 27 days ago, 2 days ago, and today');

  // Now add again 6 days from now, so that the twoDaysAgo cycle is removed from hours (which processes only 7 days)
  // and twenty seven is removed from processing 30 days
  var sixDays = addDays(now, 6);
  const sixDaysCycle = fakeCycle(sixDays, 12, "part111", 1, 'palcc');
  st = events.reducer(
    st,
    evtsToAction(sixDays, sixDaysCycle)
  );

  // twentySevenDaysAgo should have been filtered out

  expect(st.last30).toMatchSnapshot('last30 with 2 days ago, today, and 6 days from now');

  // empty list should keep lists unchanged and the same object
  let newSt = events.reducer(
    st,
    {
      type: events.ActionType.LoadRecentLogEntries,
      now: sixDays,
      pledge: {
        status: PledgeStatus.Completed,
        result: []
      }
    }
  );
  expect(newSt.last30).toBe(st.last30);
}

it('processes new log events into last30', () => {
  procNewEvents((now, evts) =>
    ({
      type: events.ActionType.LoadRecentLogEntries,
      now,
      pledge: {
        status: PledgeStatus.Completed,
        result: evts
      }
    }));
});

it("refreshes new events into last30", () => {
  procNewEvents((now, evts) =>
    ({
      type: events.ActionType.ReceiveNewLogEntries,
      now,
      events: evts
    }));
});

it("starts loading a specific month for analysis", () => {
  let st = events.reducer(
    {...events.initial},
    {
      type: events.ActionType.LoadSpecificMonthLogEntries,
      month: new Date(2018, 2, 1),
      pledge: {
        status: PledgeStatus.Starting
      }
    }
  );
  expect(st.analysis_period).toBe(events.AnalysisPeriod.SpecificMonth);
  expect(st.analysis_period_month).toEqual(new Date(2018, 2, 1));
  expect(st.loading_analysis_month_log).toBe(true);
  expect(st.loading_analysis_month_jobs).toBe(false);
  expect(st.selected_month).toBe(events.initial.selected_month);
});

it("starts loading a specific month jobs for analysis", () => {
  let st = events.reducer(
    {...events.initial},
    {
      type: events.ActionType.LoadSpecificMonthJobHistory,
      month: new Date(2018, 2, 1),
      pledge: {
        status: PledgeStatus.Starting
      }
    }
  );
  expect(st.loading_analysis_month_log).toBe(false);
  expect(st.loading_analysis_month_jobs).toBe(true);
});

it("loads 30 days for analysis", () => {
  const cycles = stationCycles.process_events(
    {type: stationCycles.ExpireOldDataType.NoExpire},
    fakeCycle(new Date(), 3),
    stationCycles.initial);

  let st = events.reducer(
    {...events.initial,
      analysis_period: events.AnalysisPeriod.SpecificMonth,
      selected_month: {cycles}
    },
    {
      type: events.ActionType.SetAnalysisLast30Days
    }
  );
  expect(st.analysis_period).toBe(events.AnalysisPeriod.Last30Days);
  expect(st.selected_month).toBe(events.initial.selected_month);
});

it("loads a specific month for analysis", () => {
  var now = new Date(2018, 1, 2, 3, 4, 5);

  // start with cycles from 27 days ago, 2 days ago, and today
  const todayCycle = fakeCycle(now, 30, "partAAA", 1, 'palss');
  var twoDaysAgo = addDays(now, -2);
  const twoDaysAgoCycle = fakeCycle(twoDaysAgo, 24, "partBBB", 2, 'paltt');
  var twentySevenDaysAgo = addDays(now, -27);
  const twentySevenCycle = fakeCycle(twentySevenDaysAgo, 18, "partCCC", 3, 'paltt');

  let st = events.reducer(
    {...events.initial,
      loading_analysis_month_log: true,
      analysis_period_month: new Date(2018, 1, 1),
      analysis_period: events.AnalysisPeriod.SpecificMonth
    },
    {
      type: events.ActionType.LoadSpecificMonthLogEntries,
      month: new Date(2018, 1, 1),
      pledge: {
        status: PledgeStatus.Completed,
        result: twentySevenCycle.concat(twoDaysAgoCycle, todayCycle)
      }
    }
  );
  expect(st.analysis_period).toBe(events.AnalysisPeriod.SpecificMonth);
  expect(st.loading_analysis_month_log).toBe(false);
  expect(st.analysis_period_month).toEqual(new Date(2018, 1, 1));
  expect(st.selected_month).toMatchSnapshot('selected month with 27 days ago, 2 days ago, and today');
});

it("bins actual cycles by day", () => {
  const now = new Date(2018, 2, 5);

  const evts = ([] as ILogEntry[])
    .concat(
      fakeCycle(now, 30),
      fakeCycle(addHours(now, -3), 20),
      fakeCycle(addHours(now, -15), 15),
    );
  const st = events.reducer(
    events.initial,
    {
      type: events.ActionType.LoadRecentLogEntries,
      now: addDays(now, 1),
      pledge: {
        status: PledgeStatus.Completed,
        result: evts
      }
    });

  const byDay = events.binCyclesByDay(
    st.last30.cycles.by_part_then_stat,
    c => duration(c.active).asMinutes()
  );

  expect(byDay).toMatchSnapshot("cycles binned by day and station");
});