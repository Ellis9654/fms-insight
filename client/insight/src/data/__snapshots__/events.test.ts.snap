// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`last30 with 2 days ago, today, and 6 days from now 1`] = `
Object {
  "cycles": Object {
    "by_part_then_stat": Immutable.Map {
      part222-1: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-01-31T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-01-31T09:36:05.000Z,
                    "y": 24,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-01-31T09:46:05.000Z,
                    "y": 3,
                  },
                ],
      },
      part111-1: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-02-02T09:09:05.000Z,
                    "y": 6,
                  },
                  Object {
                    "x": 2018-02-08T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-02-02T09:42:05.000Z,
                    "y": 30,
                  },
                  Object {
                    "x": 2018-02-08T09:24:05.000Z,
                    "y": 12,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-02-02T09:52:05.000Z,
                    "y": 3,
                  },
                  Object {
                    "x": 2018-02-08T09:34:05.000Z,
                    "y": 3,
                  },
                ],
      },
},
  },
  "latest_counter": 5,
  "latest_event": 2018-01-06T09:40:05.000Z,
  "oee": Object {
    "last_week_of_hours": Immutable.List [
      Object {
            "date": 2018-02-02T09:09:05.000Z,
            "hours": 0.1,
            "station": "L/U #1",
          },
      Object {
            "date": 2018-02-02T09:42:05.000Z,
            "hours": 0.5,
            "station": "MC #1",
          },
      Object {
            "date": 2018-02-02T09:52:05.000Z,
            "hours": 0.05,
            "station": "L/U #2",
          },
      Object {
            "date": 2018-02-08T09:09:05.000Z,
            "hours": 0.1,
            "station": "L/U #1",
          },
      Object {
            "date": 2018-02-08T09:24:05.000Z,
            "hours": 0.2,
            "station": "MC #1",
          },
      Object {
            "date": 2018-02-08T09:34:05.000Z,
            "hours": 0.05,
            "station": "L/U #2",
          },
],
    "system_active_hours_per_week": 168,
  },
}
`;

exports[`last30 with 27 days ago, 2 days ago, and today 1`] = `
Object {
  "cycles": Object {
    "by_part_then_stat": Immutable.Map {
      part222-2: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-01-06T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-01-06T09:30:05.000Z,
                    "y": 18,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-01-06T09:40:05.000Z,
                    "y": 3,
                  },
                ],
      },
      part222-1: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-01-31T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-01-31T09:36:05.000Z,
                    "y": 24,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-01-31T09:46:05.000Z,
                    "y": 3,
                  },
                ],
      },
      part111-1: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-02-02T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-02-02T09:42:05.000Z,
                    "y": 30,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-02-02T09:52:05.000Z,
                    "y": 3,
                  },
                ],
      },
},
  },
  "latest_counter": 5,
  "latest_event": 2018-01-06T09:40:05.000Z,
  "oee": Object {
    "last_week_of_hours": Immutable.List [
      Object {
            "date": 2018-01-31T09:09:05.000Z,
            "hours": 0.1,
            "station": "L/U #1",
          },
      Object {
            "date": 2018-01-31T09:36:05.000Z,
            "hours": 0.4,
            "station": "MC #1",
          },
      Object {
            "date": 2018-01-31T09:46:05.000Z,
            "hours": 0.05,
            "station": "L/U #2",
          },
      Object {
            "date": 2018-02-02T09:09:05.000Z,
            "hours": 0.1,
            "station": "L/U #1",
          },
      Object {
            "date": 2018-02-02T09:42:05.000Z,
            "hours": 0.5,
            "station": "MC #1",
          },
      Object {
            "date": 2018-02-02T09:52:05.000Z,
            "hours": 0.05,
            "station": "L/U #2",
          },
],
    "system_active_hours_per_week": 168,
  },
}
`;

exports[`selected month with 27 days ago, 2 days ago, and today 1`] = `
Object {
  "cycles": Object {
    "by_part_then_stat": Immutable.Map {
      partCCC-3: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-01-06T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-01-06T09:30:05.000Z,
                    "y": 18,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-01-06T09:40:05.000Z,
                    "y": 3,
                  },
                ],
      },
      partBBB-2: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-01-31T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-01-31T09:36:05.000Z,
                    "y": 24,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-01-31T09:46:05.000Z,
                    "y": 3,
                  },
                ],
      },
      partAAA-1: Immutable.Map {
            L/U #1: Array [
                  Object {
                    "x": 2018-02-02T09:09:05.000Z,
                    "y": 6,
                  },
                ],
            MC #1: Array [
                  Object {
                    "x": 2018-02-02T09:42:05.000Z,
                    "y": 30,
                  },
                ],
            L/U #2: Array [
                  Object {
                    "x": 2018-02-02T09:52:05.000Z,
                    "y": 3,
                  },
                ],
      },
},
  },
}
`;
